core/sequences/sequences.factor

```diff
@@ -31,9 +31,8 @@ M: sequence shorten 2dup length < [ set-length ] [ 2drop ] if ; inline
 : if-empty ( ..a seq quot1: ( ..a -- ..b ) quot2: ( ..a seq -- ..b ) -- ..b )
     [ dup empty? ] [ [ drop ] prepose ] [ ] tri* if ; inline
 
-: when-empty ( seq quot -- ) [ ] if-empty ; inline
-
-: unless-empty ( seq quot -- ) [ ] swap if-empty ; inline
+: when-empty ( ... seq quot: ( ... -- seq ) -- ... seq ) [ ] if-empty ; inline
```

Comment:
The stack effect for `when-empty` implies that it's the same `seq` that gets output but that's not the case since it's dropped.

What about:
```
: when-empty ( ... seq quot: ( ... -- obj ) -- ... obj ) [ ] if-empty ; inline
```

Reply:
or `seq'` if we assume these output a seq

Reply:
This implies that the output is always the result of the quotation, but in some cases it'll be the original `seq`.

Reply:
The following was suggested in Discord
`: when-empty ( ... seq quot: ( ... -- v ) -- ... seq/v ) [ ] if-empty ; inline`

Reply:
How about this as a compromise?
`: when-empty ( ... seq quot: ( ... -- seq' ) -- ... seq/' ) [ ] if-empty ; inline`

Reply:
Or maybe `: when-empty ( ... seq quot: ( ... -- obj ) -- ... seq/obj ) [ ] if-empty ; inline`?

Reply:
`seq/obj` is the best so far?

Reply:
Pushed.